validate/src/command.rs

```diff
@@ -0,0 +1,277 @@
+
+                f.write_str(" characters")
+            }
+            CommandValidationErrorType::PermissionsCountInvalid => {
+                f.write_str("more than ")?;
+                Display::fmt(&GUILD_COMMAND_PERMISSION_LIMIT, f)?;
+
+                f.write_str(" permission overwrites were set")
+            }
+        }
+    }
+}
+
+impl Error for CommandValidationError {}
+
+/// Type of [`CommandValidationError`] that occurred.
+#[derive(Debug)]
+#[non_exhaustive]
+pub enum CommandValidationErrorType {
+    /// Too many commands have been provided.
+    ///
+    /// The maximum number of commands is defined by
+    /// [`GUILD_COMMAND_LIMIT`].
+    CommandCountInvalid,
+    /// Required command options have to be passed before optional ones.
+    CommandOptionsRequiredFirst {
+        /// Index of the option that failed validation.
+        index: usize,
+    },
+    /// Command description is invalid.
+    DescriptionInvalid,
+    /// Command name is invalid.
+    NameInvalid,
+    /// More than 10 permission overwrites were set.
+    PermissionsCountInvalid,
+}
+
+/// Validate a [`Command`].
+///
+/// # Errors
+///
+/// Returns an error with type [`DescriptionInvalid`] if the description is
+/// invalid.
+///
+/// Returns an error with type [`NameInvalid`] if the name is invalid.
+///
+/// [`DescriptionInvalid`]: CommandValidationErrorType::DescriptionInvalid
+/// [`NameInvalid`]: CommandValidationErrorType::NameInvalid
+pub fn command(value: &Command) -> Result<(), CommandValidationError> {
+    let Command {
+        description, name, ..
+    } = value;
+
+    self::description(description)?;
+
+    self::name(name)?;
+
+    Ok(())
+}
+
+/// Validate the description of a [`Command`].
+///
+/// The length of the description must be more than [`DESCRIPTION_LENGTH_MIN`]
+/// and less than or equal to [`DESCRIPTION_LENGTH_MAX`].
+///
+/// # Errors
+///
+/// Returns an error with type [`DescriptionInvalid`] if the description is
+/// invalid.
+///
+/// [`DescriptionInvalid`]: CommandValidationErrorType::DescriptionInvalid
+pub fn description(value: impl AsRef<str>) -> Result<(), CommandValidationError> {
+    let len = value.as_ref().chars().count();
+
+    // https://discord.com/developers/docs/interactions/application-commands#application-command-object-application-command-option-structure
+    if (DESCRIPTION_LENGTH_MIN..=DESCRIPTION_LENGTH_MAX).contains(&len) {
+        Ok(())
+    } else {
+        Err(CommandValidationError {
+            kind: CommandValidationErrorType::DescriptionInvalid,
+        })
+    }
+}
+
+/// Validate the name of a [`Command`].
+///
+/// The length of the name must be more than [`NAME_LENGTH_MIN`] and less than
+/// or equal to [`NAME_LENGTH_MAX`].
+///
+/// # Errors
+///
+/// Returns an error with type [`NameInvalid`] if the name is invalid.
+///
+/// [`NameInvalid`]: CommandValidationErrorType::NameInvalid
+pub fn name(value: impl AsRef<str>) -> Result<(), CommandValidationError> {
```

Comment:
I think we decided initially that due to command names possibly being in non-Latin script languages, it is up to the user to provide a lowercase name. Regardless, it could be documented that our validation specifically does not do that.

Reply:
[`char::to_lowercase`](https://doc.rust-lang.org/std/primitive.char.html#method.to_lowercase) use the Unicode Character Database and works with non-Latin script languages. 

[Playground example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=dda7c40481c5aef75e0c8f5410b8e478)

Reply:
I'd like to merge this PR for now, but if you would like to make an issue for this I would be happy to get to it at some point before the next major release.

Reply:
Opened a PR for that: #1395